File: Perl/IPC_HW/hw5.pl
#!/usr/bin/perl

sub judge {
    return 0 if $_[0] eq '+' || $_[0] eq '-';
    return 1 if $_[0] eq '.';
    return 2 if ord '0' <= ord $_[0] && ord $_[0] <= ord '9';
    return 3 if $_[0] eq '\n';
}

sub f {
    $n = shift;
    $q = 's';
    for (0 .. length($n) - 1) {
        $j = judge(substr($n, $_, 1));
        if ($q eq 's') {
            $q = ($j == 0 ? 's' : $j == 1 ? 'f' : $j == 2 ? 'd' : 'e');
        }
        elsif ($q eq 'd') {
            $q = ($j == 2 ? 'd' : $j == 1 ? 'r' : 'e');
        }
        elsif ($q eq 'f') {
            $q = ($j == 2 ? 'r' : 'e');
        }
        elsif ($q eq 'r') {
            $q = ($j == 2 ? 'r' : 'e');
        }
    }
    return $q eq 'r';
}

print "Enter a real constant:";
while (chomp($_ = <>)) {
    print f($_) ? "Accepted\n" : "Rejected\n";
    print "Enter a real constant:";
}

File: Python/IPC_HW/hw5.py
#!/usr/bin/env python

import sys

def judge (n) :
    if n == '+' or n == '-' :
        return 0
    elif n == '.' :
        return 1
    elif '0' <= n <= '9' :
        return 2
    elif n == '\n' :
        return 3

def f (n) :
    q = 's'
    for i in range (0, len(n)) :
        j = judge (n[i:i+1])
        if q == 's' :
            q = ('s' if j == 0 else 'd' if j == 2 else 'f' if j == 1 else 'e')
        elif q == 'd' :
            q = ('d' if j == 2 else 'r' if j == 1 else 'e')
        elif q == 'f' :
            q = ('r' if j == 2 else 'e')
        elif q == 'r' :
            q = ('r' if j == 2 else 'e')


    return q == 'r'

if __name__ == '__main__' :
    while 1 :
        try :
            n = raw_input ('Enter a real constant: ')
            sys.stdout.write ("%s\n" % ("Accepted" if f (n) else "Rejected"))
        except EOFError :
            break

